packages/run-protocol/src/runStake/runStakeManager.js

```diff
@@ -5,123 +5,156 @@ import { makeRatio } from '@agoric/zoe/src/contractSupport/ratio.js';
-    const [[_addr, valueLiened]] = getCopyBagEntries(attestationGiven.value);
-    const amountLiened = AmountMath.make(brands.Stake, valueLiened);
-    const maxDebt = floorMultiplyBy(amountLiened, mintingRatio);
-    return { maxDebt, amountLiened };
-  };
+  const totalDebt = AmountMath.makeEmpty(brands.debt, 'nat');
+  const compoundedInterest = makeRatio(100n, brands.debt); // starts at 1.0, no interest
+  const latestInterestUpdate = startTimeStamp;
 
-  let totalDebt = AmountMath.makeEmpty(brands.debt, 'nat');
-  let compoundedInterest = makeRatio(100n, brands.debt); // starts at 1.0, no interest
-  let latestInterestUpdate = startTimeStamp;
-  const { updater: assetUpdater, notifier: assetNotifer } = makeNotifierKit(
+  const { updater: assetUpdater, notifier: assetNotifier } = makeNotifierKit(
     harden({
       compoundedInterest,
-      interestRate: paramManager.getInterestRate(),
+      interestRate: mintPowers.getGovernedParams().getInterestRate(),
       latestInterestUpdate,
       totalDebt,
     }),
   );
 
+  // ??? does this promise need to be a presence?
+  const periodNotifier = E(timerService).makeNotifier(0n, recordingPeriod);
+  const { zcfSeat: poolIncrementSeat } = zcf.makeEmptySeatKit();
+
+  return {
+    assetNotifier,
+    assetUpdater,
+    brands,
+    chargingPeriod,
+    compoundedInterest,
+    debtMint,
+    latestInterestUpdate,
+    mintPowers,
+    periodNotifier,
+    poolIncrementSeat,
+    recordingPeriod,
+    startTimeStamp,
+    totalDebt,
+    zcf,
+  };
+};
+
+/**
+ *
+ * @param {MethodContext} context
+ */
+const finish = ({ state }) => {
+  const { debtMint, mintPowers, periodNotifier, poolIncrementSeat, zcf } =
+    state;
+
   /**
    *
    * @param {bigint} updateTime
-   * @param {ZCFSeat} poolIncrementSeat
    */
-  const chargeAllVaults = async (updateTime, poolIncrementSeat) => {
+  const chargeAllVaults = async updateTime => {
     trace('chargeAllVaults', { updateTime });
-    const interestRate = paramManager.getInterestRate();
+    const interestRate = mintPowers.getGovernedParams().getInterestRate();
 
-    // Update local state with the results of charging interest
-    ({ compoundedInterest, latestInterestUpdate, totalDebt } = chargeInterest(
+    const changes = chargeInterest(
       {
         mint: debtMint,
-        mintAndReallocateWithFee,
+        mintAndReallocateWithFee: mintPowers.mintAndReallocate,
         poolIncrementSeat,
         seatAllocationKeyword: KW.Debt,
       },
       {
         interestRate,
-        chargingPeriod,
-        recordingPeriod,
+        chargingPeriod: state.chargingPeriod,
+        recordingPeriod: state.recordingPeriod,
+      },
+      {
+        latestInterestUpdate: state.latestInterestUpdate,
+        compoundedInterest: state.compoundedInterest,
+        totalDebt: state.totalDebt,
       },
-      { latestInterestUpdate, compoundedInterest, totalDebt },
       updateTime,
-    ));
+    );
+    Object.assign(state, changes);
```

Comment:
Hm. Is this `Object.assign(...)` a recommended pattern? It throws type safety right out the window, no?

Reply:
Good point! WDYT of a [helper for type safety](https://www.typescriptlang.org/play?filetype=js#code/LAKA9AVBoAQTABALgUwLYAcA2BDVMAVGAMwFcstZ4EMcAnHNGAbwIF8TzKQ5FaGmzAAr0kASxxYAPAQB8HfuMlUwoAMYB7AHYBnJDEUSsAQR06xAcy0wAvDAAUZCgBoDoowEpbslrBgB5ACMAKxQ1JAA6HDNLLUcuV0NJDwBuUDY0kFB1bT0YPTwUWxYtRhQALhgAclQAEyrXTSwNOkqquhR6jOyQJJMYq3sC1FdmUrQK6s6qtlSevtNzQeGUUYBrFABPNrqZuaze90lF2KGkQtHxyYBmWcygA)?